src/renderer/dx/DxSoftFont.cpp

```diff
@@ -0,0 +1,256 @@
+    // code above, we also need to pad the top and bottom. But in this case we
+    // can simply do a whole row of glyphs from the grid at the same time.
+    for (auto gridRow = 0u; gridRow < BITMAP_GRID_HEIGHT; gridRow++)
+    {
+        const auto rowOffset = _yOffsetForGlyph<size_t>(gridRow);
+        const auto rowTop = bitmapScanline(rowOffset);
+        const auto rowBottom = bitmapScanline(rowOffset + _sourceSize.height - 1);
+        for (auto i = 1; i <= PADDING; i++)
+        {
+            std::copy_n(rowTop, bitmapWidth, rowTop - i * bitmapWidth);
+            std::copy_n(rowBottom, bitmapWidth, rowBottom + i * bitmapWidth);
+        }
+    }
+}
+
+HRESULT DxSoftFont::SetTargetSize(const til::size targetSize)
+{
+    _targetSize = targetSize;
+    return _scaleEffect ? _scaleEffect->SetValue(D2D1_SCALE_PROP_SCALE, _scaleForTargetSize()) : S_OK;
+}
+
+HRESULT DxSoftFont::SetAntialiasing(const bool antialiased)
+{
+    _interpolation = (antialiased ? ANTIALIASED_INTERPOLATION : ALIASED_INTERPOLATION);
+    return _scaleEffect ? _scaleEffect->SetValue(D2D1_SCALE_PROP_INTERPOLATION_MODE, _interpolation) : S_OK;
+}
+
+HRESULT DxSoftFont::SetColor(const D2D1_COLOR_F& color)
+{
+    // Since our source image is monochrome, we don't care about the
+    // individual color components. We just multiply the red component
+    // by the active color value to get the output color. And note that
+    // the alpha matrix entry is already set to 1 in the constructor,
+    // so we don't need to keep updating it here.
+    _colorMatrix.m[0][0] = color.r;
+    _colorMatrix.m[0][1] = color.g;
+    _colorMatrix.m[0][2] = color.b;
+    return _colorEffect ? _colorEffect->SetValue(D2D1_COLORMATRIX_PROP_COLOR_MATRIX, _colorMatrix) : S_OK;
+}
+
+HRESULT DxSoftFont::Draw(const DrawingContext& drawingContext,
+                         const gsl::span<const Cluster> clusters,
+                         const float originX,
+                         const float originY)
+{
+    ComPtr<ID2D1DeviceContext> d2dContext;
+    RETURN_IF_FAILED(drawingContext.renderTarget->QueryInterface(d2dContext.GetAddressOf()));
+
+    // We start by creating a clipping rectangle for the region we're going to
+    // draw, and this is initially filled with the active background color.
+    D2D1_RECT_F rect;
+    rect.top = originY + drawingContext.topClipOffset;
+    rect.bottom = originY + _targetSize.height - drawingContext.bottomClipOffset;
+    rect.left = originX;
+    rect.right = originX + _targetSize.width * clusters.size();
+    d2dContext->FillRectangle(rect, drawingContext.backgroundBrush);
+    d2dContext->PushAxisAlignedClip(rect, D2D1_ANTIALIAS_MODE_ALIASED);
+    auto resetClippingRect = wil::scope_exit([&]() noexcept { d2dContext->PopAxisAlignedClip(); });
+
+    // The bitmap and associated scaling/coloring effects are created on demand
+    // so we need make sure they're generated now.
+    RETURN_IF_FAILED(_createResources(d2dContext.Get()));

```

Comment:
I haven't made it all the way through, but how do we handle device invalidation? Are the bitmaps okay to use cross-Context if we end up having to regenerate the context in the main renderer?

Reply:
Nevermind. `Reset` very obviously does that, and it happens to be called from `releaseDeviceResources`. Derp. Thanks.